Library/Homebrew/rubocops/cask/no_overrides.rb

```diff
@@ -0,0 +1,60 @@
+# typed: true
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Cask
+      class NoOverrides < Base
+        extend T::Sig
+        include CaskHelp
+
+        ON_SYSTEM_METHODS = RuboCop::Cask::Constants::ON_SYSTEM_METHODS
+        MESSAGE = <<~EOS
+          Do not use a top-level `%<stanza>s` stanza as the default. Add it to an `on_{system}` block instead.
+          Use `:or_older` or `:or_newer` to specify a range of macOS versions.
+        EOS
+
+        def on_cask(cask_block)
+          cask_stanzas = cask_block.toplevel_stanzas
+
+          # Skip if there are no `on_*` blocks.
+          return unless (on_blocks = cask_stanzas.select { |s| ON_SYSTEM_METHODS.include?(s.stanza_name) }).any?
+
+          stanzas_in_blocks = on_system_stanzas(on_blocks)
+
+          cask_stanzas.each do |stanza|
+            # Skip if the stanza is itself an `on_*` block.
+            next if ON_SYSTEM_METHODS.include?(stanza.stanza_name)
+            # Skip if the stanza outside of a block is not also in an `on_*` block.
+            next unless stanzas_in_blocks.include?(stanza.stanza_name)
+
+            add_offense(stanza.source_range, message: format(MESSAGE, stanza: stanza.stanza_name))
+          end
+        end
+
+        def on_system_stanzas(on_system)
+          names = Set.new
+          method_nodes = on_system.map(&:method_node)
+          method_nodes.each do |node|
+            next unless node.block_type?
+
+            node.child_nodes.each do |child|
+              child.each_node(:send) do |send_node|
+                # Skip (nested) livecheck blocks as its `url` is different to a download `url`.
+                next if send_node.method_name == :livecheck
+                next if send_node.parent.block_type? && send_node.parent.method_name == :livecheck
```

Comment:
I think we should have a helper function for this. This is the same check that is needed in `toplevel_stanzas`, but for `livecheck` instead of `cask`, right?